fun.lua

```diff
@@ -1032,6 +1040,86 @@ methods.operator = operator
 exports.op = operator
 methods.op = operator
 
+--------------------------------------------------------------------------------
+-- Stateful iterators
+--------------------------------------------------------------------------------
+
+-- Returns 'gen' throwing an error that explains that method is not supported
+local unsupported_stateful_method_gen = function(method)
+    return function()
+        error(string.format("stateful iterators do not support method '%s'",
+              method))
+    end
+end
+
+-- Helper of 'wrap_with_elems'
+local gen_with_elems_wrapper = function(gen_param_elems, state)
+    local elems = gen_param_elems[3]
+    if not elems then
+        return gen_param_elems[1](gen_param_elems[2], state)
+    else
+        gen_param_elems[3] = nil
+        return state, unpack(elems)
+    end
+end
+
+-- Wrap an iterator that yields all values from 'elems' table on the first
+-- iteration and then works in the same way as 'wrap(gen, param, state)'
+local wrap_with_elems = function(elems, gen, param, state)
+    local gen_param_elems = {gen, param, elems}
+    return wrap(gen_with_elems_wrapper, gen_param_elems, state)
+end
+
+-- Checks if drop_while should continue skipping. If iterator is not exhausted,
+-- elements returned by iterator are wrapped into a table and returned as the
+-- third return value
+local stateful_drop_while_check = function(fun, state_x, ...)
+    if state_x == nil then
+        return state_x, false
+    end
+    return state_x, fun(...), {...}
+end
+
+-- Implementation of method drop_while: original one copies iterator state to
+-- iterate over the same element twice - it doesn't work with stateful iterators
+local stateful_drop_while = function(fun, gen_x, param_x, state_x)
```

Comment:
I've experimented with this code for some time and ends up with some alternative implementation.

```lua
-- Creates a singleton iterator: it returns the given values on
-- the first iteration and ends then.
local repeat_once = function(...)
    return take_n(1, duplicate(...))
end

-- See stateful_drop_while().
local function stateful_drop_while_impl(fun, gen, param, state, ...)
    -- If the iterator is exhausted, return nil iterator.
    if state == nil then
        return wrap(nil_gen, nil, nil)
    end
    -- While the predicate is positive, advance the iterator.
    if fun(...) then
        return stateful_drop_while_impl(fun, gen, param, gen(param, state))
    end
    -- Once the predicate becomes negative, return an iterator
    -- that repeats the last value and then returns all the
    -- following ones from the original iterator.
    return chain(repeat_once(...), wrap(gen, param, state))
end

-- Advances the given iterator until the function becomes false,
-- then returns the current element on the first iteration and
-- the remaining elements on the next iterations.
--
-- Unlike drop_while() this implementation doesn't attempt to
-- repeat a previous iteration by calling the `gen` function with
-- the previous `state`.
local stateful_drop_while = function(fun, gen, param, state)
    assert(type(fun) == "function", "invalid first argument to drop_while")
    return stateful_drop_while_impl(fun, gen, param, gen(param, state))
end
```

It looks simpler for my taste and I see no obvious drawbacks. Do you see any?

I hope that reusing existing luafun methods is a good direction to make the code more maintainable.

It requires a fixup in the `chain` method:

```diff
diff --git a/fun.lua b/fun.lua
index 627c652..6023ab9 100644
--- a/fun.lua
+++ b/fun.lua
@@ -951,7 +951,7 @@ local chain_gen_r2 = function(param, state, state_x, ...)
     if state_x == nil then
         local i = state[1]
         i = i + 1
-        if param[3 * i - 1] == nil then
+        if param[3 * i - 2] == nil then
             return nil
         end
         local state_x = param[3 * i]
```

Before this fixup `chain` assumes that there are no next iterator if its `param` is `nil`. The fix replaced it with checking of the `gen` function.

Reply:
Performance measurements have shown that an implementation using `chain` is pretty slower. Especially with JIT disabled. Moreover, `chain` modifies state (for example, it modifies indexes of elements when using `pairs(table)`) - seems to be a breaking change. So it was decided to stick with initial implementation.

Reply:
OK. Thanks for looking into it!

The only remark here that I would like to have a test case that verifies that the indexes (state values) are unmodified.

Reply:
Introduced such tests for all filters and slicers in a separate commit.

Reply:
Nice!

Reply:
> It requires a fixup in the `chain` method:
>
> ```diff
> diff --git a/fun.lua b/fun.lua
> index 627c652..6023ab9 100644
> --- a/fun.lua
> +++ b/fun.lua
> @@ -951,7 +951,7 @@ local chain_gen_r2 = function(param, state, state_x, ...)
>      if state_x == nil then
>          local i = state[1]
>          i = i + 1
> -        if param[3 * i - 1] == nil then
> +        if param[3 * i - 2] == nil then
>              return nil
>          end
>          local state_x = param[3 * i]
> ```
>
> Before this fixup `chain` assumes that there are no next iterator if its `param` is `nil`. The fix replaced it with checking of the `gen` function.

Filed #86 about it.